### DISCUSSION TITLE
[Policy] About Forking and Promotion of Forks

### DISCUSSION THREAD

**Discussion by lllyasviel** [lllyasviel/Fooocus#699] [Announcements]
Recently the community of Fooocus has witnessed a special development mode of “forks”. Forks are helpful for distinguishing whether any features are really necessary, and whether any features are just like pre-installed adware/malware that people want to uninstall immediately after they buy a new Android smartphone.

This development mode may have already influenced many other successful and larger projects like Android’s AOSP, Chrome’s Chromium, etc.

We want to stimulate the trend and add the below policy

1. Forking is encouraged. After you fork Fooocus, you can report here for a promotion – your link will be added to the main [readme’s “Advanced” section](https://github.com/lllyasviel/Fooocus#advanced-features), and the ending of [Fooocus Advanced](https://github.com/lllyasviel/Fooocus/discussions/117#forks), and this post, so that we can maximize the exposure of your work. We will also consider adding some other places in the future.
2. Forks (that have been announced) will be observed by our team at a monthly or half-monthly period. Features will be reviewed and considered by Fooocus if they have been proved to be really necessary and really get popular with a large enough group of people.
3. Discussions of forks are also encouraged when the topic is fork-related, but spamming is prohibited (for example, spamming fork’s link by replying every message/post/issue in this repo). Challenging this may result in account being banned.
4. “Do not be a bad ass.” Attacking, insulting, and discriminations are wrong. Challenging this may result in account being banned.

Thank you for your support to Fooocus as always.

**Comment by lllyasviel** [comment]
Some existing forks that I know:

[runew0lf/RuinedFooocus](https://github.com/runew0lf/RuinedFooocus)

[MoonRide303/Fooocus-MRE](https://github.com/MoonRide303/Fooocus-MRE)

**Reply by 3Diva** [reply]
I wish there was a RuinedFooocus "Discussions" board. I don't want to try it out if there's no place to leave feedback or ask questions for that fork. 

Might be for the best, I am already really enjoying Fooocus and Fooocus-MRE! :)

**Reply by MoonRide303** [reply]
@3Diva Time to time I talk with @runew0lf on Discord, and he runs [Discord server](https://github.com/runew0lf/RuinedFooocus#discord-support-server) to allow direct communication with RuinedFooocus users.

@lllyasviel maybe you could create official chat server (Discord / Matrix / whatever), where all Fooocus (either original, or forks) users and contributors could communicate with each other more directly? Both A1111 (on Discord) and ComfyUI (on Matrix) have such as spaces, and it might be more convenient than GitHub discussions.

**Reply by 3Diva** [reply]
Thank you, @MoonRide303 - I don't use Discord though. It would be nice if he had a Discussions section like both you and @lllyasviel have. It's hugely helpful when seeking answers to questions or just giving feedback. It's a real shame he doesn't have that too.

**Reply by runew0lf** [reply]
i just added one :D

**Comment by lllyasviel** [comment]
Newly added fenneishi/Fooocus-Control: https://github.com/fenneishi/Fooocus-Control

**Reply by fenneishi** [reply]
thanks :D

**Comment by tiwentichat** [comment]
> Newly added fenneishi/Fooocus-Control: https://github.com/fenneishi/Fooocus-Control



> Newly added fenneishi/Fooocus-Control: https://github.com/fenneishi/Fooocus-Control

This project is also very cool!

**Reply by PeiPeiWow** [reply]
Yeah, so cool!!!!!

**Reply by fenneishi** [reply]
🎉 🎉 🎉

**Reply by 3Diva** [reply]
Can someone let me know how Fooocus-Control differs from Fooocus-MRE? My hard drive space is getting pretty small and so I'd rather not download Fooocus-Control if it can't do something better than Fooocus-MRE already can. If someone with familiarity with Fooocus-MRE and Fooocus-Control could let me know if "Control" can do more things than "MRE" can, I'd very much appreciate it! :D

**Reply by MoonRide303** [reply]
@fenneishi I am slowly reviewing and merging MRE features back into the core (original / vanilla Fooocus), as in the long-term I think it's better approach if we have have single high quality code base, with all the features that are clear value added for both first-time and advanced users available for all of the Fooocus fans. This approach also makes fork maintenance a lot easier (you don't have to resolve so many conflicts when delta between two code bases is smaller).

As of Fooocus-Control I took a quick look on the source, and it started as copy of the original/vanilla repository (not MRE), so I wouldn't call it successor of MRE - it's your own fork done your way. I've notice for example you've added support for Recolor - I could easily add it into MRE, but I did some tests with various workflows and wasn't happy with the quality of results - so I didn't include it.

**Reply by fenneishi** [reply]
I apologize for my inappropriate expression. I have no intention of belittling Fooocus-MRE. What I mean to convey is that Fooocus-Control want to follow the work of Fooocus-MRE and hopes to become its successor. To provide a concise response for 3Diva, I didn't elaborate further on the plans and ideas for Fooocus-MRE. I will be more careful about this in the future (I believe it was necessary since I mentioned Fooocus-MRE).

Fooocus-MRE is really impressive. To be honest, I am a  big fan of both Fooocus and Fooocus-MRE. Initially, I used deep image control directly from Fooocus-MRE to generate images. However, I noticed that Fooocus 2.1 had been upgraded, and there were no plans to update Fooocus-MRE (I carefully read your instructions, and you believe that contributing directly to the original Fooocus is a more effective way). So, I thought I could develop it myself.

I strongly agree with your idea that 'as in the long-term I think it's a better approach if we have a single high-quality code base.' However, because 'lllyasviel' believes that forks are helpful for distinguishing whether any features are really necessary, I thought open-sourcing Fooocus-Control might have some merit. So I  open source it.

**Reply by fenneishi** [reply]
Based on the above discussion, I just wanted to  to explain why I added the 'ReColor ' feature:

The “ReColor” effect is really not that impressive, I added this feature because I wanted to give this feature **_A Starting Point:_**

**_Present a feature on the product. If many people need this feature and this feature is not ideal, then try to improve this feature, otherwise delete it._**

If the feature is not shown, users may not know about the feature, and it is difficult for developers to know what they need pay more attentions to optimize. In my development experience, information asymmetry between developers and users is often one of the stumbling blocks to product success.

**Of course, this must all based on continuing to maintain one of the core features of fooocus -- simplicity(manual tweaking is not needed, and users only need to focus on the prompts and images.)**. so I probably wouldn't add a 'start at' left to a 'stop at' because that would put too much unnecessary mental burden on the user. But I think adding a image control in Advanced UI is perfectly acceptable for complexity to increase.

In addition, (as lllyasviel said:Forks are helpful for distinguishing whether any features are really necessary, and whether any features are just like pre-installed adware/malware that people want to uninstall immediately after they buy a new Android smartphone.) .I feel that the various fork projects based on fooocus, which are somewhat similar to the multiverse, experimental version of fooocus, are a very effective way to find a balance between simplicity and variety. And users can meet their own diverse needs. So in fact, a variety of fork projects is one of the important ways to keep fooocus simple.

**Comment by pppoe** [comment]
Thanks for the nice work! I enjoyed it a lot. I have a version here with Efficient SAM added for auto mask generation under the inpainting tab 
https://github.com/pppoe/Fooocus-SAM

Blog post at https://blog.haoxiang.org/2024/01/added-sam-to-fooocus/

**Reply by llinlluo** [reply]
How to install?tks.

**Reply by pppoe** [reply]
- generally follow [FOOOCUS README file](https://github.com/pppoe/Fooocus-SAM/blob/main/readme.md) to setup python3
- ``python3 launch.py`` will install deps, download models and launch Gradio

**Reply by mashb1t** [reply]
@pppoe will check out EfficientSAM, currently experimenting with SAM and GroundingDINO in general for better mask generation and automated multi-step optimizations.

- discussion in https://github.com/lllyasviel/Fooocus/discussions/3122
- changes in https://github.com/mashb1t/Fooocus/compare/main...mashb1t:Fooocus:feature/multi-step-rendering

**Comment by tiwentichat** [comment]
> Thanks for the nice work! I enjoyed it a lot. I have a version here with Efficient SAM added for auto mask generation under the inpainting tab https://github.com/pppoe/Fooocus-SAM
> 
> Blog post at https://blog.haoxiang.org/2024/01/added-sam-to-fooocus/

Very cool feather, will you push this to Fooocus main branch or develop it independently?

**Reply by pppoe** [reply]
Thanks! I am happy to submit a PR but as per the rules above the team will decide.

"Forks (that have been announced) will be observed by our team at a monthly or half-monthly period. Features will be reviewed and considered by Fooocus if they have been proved to be really necessary and really get popular with a large enough group of people"

**Comment by dcryptoniun** [comment]


**Reply by 3Diva** [reply]
What are the differences in your "DFooocus_lightning" compared to the original Fooocus (besides just changing the model and steps to Juggernaut-XL-Lightning)?

**Comment by TanvirHafiz** [comment]
why dont all these forks add the faceswap feature to their UI? i keep going back to fooocus original only for this one feaure which is immensely popular and is the industry standard for virtual avatar creation now!

**Comment by mrhan1993** [comment]
I think this is useful, https://github.com/mrhan1993/FooocusAPI

I rewrite Fooocus-API and dev a new version based on Fooocus fork
